### PR TITLE
real DI scoping

### LINKED ISSUE #12: Scoping Context Injection to Scenario Execution Lifetimes
The default Context Injection in SpecFlow is scoped to the life-time of scenario execution, how do we go about achieving this with SpecFlow.DependencyInjection?

As steps are global across the entire project, I have separated the steps into logical units/binding classes where a single scenario may invoke steps from multiple bindings. This means I need to share the same context instance between each. Right now I'm having trouble because each binding class is receiving its own context instance. Adding the context instances as singletons won't help either.

### PR DESCRIPTION
**Pull request by shlomiassaf**
The current implementation is missing some of the goals of DI.

- It will create a new service container on every scenario (which might be slow)
  - It will build the service provider on every scenario  (which might be slow)
- The above is true for Features as well, i.e. Injections from Features will not be the same as in Scenario!!!
- It does not provide singleton option in the DI, as the DI is built every time
- It does not dispose the service providers, it's not possible since it's always the root provider
- It does not provide scope resolution. I.E Feature based scoping vs Scenario based scoping
- Because a new container was created on every scenario, a new instance of the binding class was created, not allowing property share via instance.

----

This PR fixes scoping for **Scenario** and add support for **Feature** scoping.

With **Scenario**:

- The service container is now created and built ONCE for the entire run
- Singleton are now possible for the entire run.
- Each **Scenario** is now a real `IServiceScope` which is also disposed once the **Scenario** is done

With **Feature**:

- The service container is now created and built ONCE for the entire run
- Singleton are now possible for the entire run.
- Each **Feature** is a `IServiceScope` which is also disposed once the **Feature** is done
- All Scenarios are injected using the same scope, i.e. they share injections
- **Feature** hooks can now get injections shared with the injection of the **Scenario**

To use **Feature** scoping

```
[ScenarioDependencies(ScopeLevel = ScopeLevelType.Feature)]
```

> The default is **Scenario** to ensure backward compatibility

### CODE REVIEW

**Comment by shlomiassaf** [comment]
@mbhoek I know it might look overwhelming, I hope you invest the time to get into it, because this is a big improvment.

**Comment by mbhoek** [comment]
@shlomiassaf Thank you for your contribution! I'm currently busy working on another project, but will have time to review your PR during the upcoming holidays. Sorry for the delay!

**Comment by mbhoek** [comment]
Hello @shlomiassaf, I finally found some time to review your PR and I think it's excellent work. I'm going to include in the next big update of this plugin, which will also include fixes for #75. I'd like to combine those fixes so I want to merge your work into a different branch `next` instead of `main`. I'm moving forward with this assuming you are okay with this. Thanks again!

**Comment by shlomiassaf** [comment]
I appreciate your update.
Looking forward for it, we're working with a side branch.

**Comment by mbhoek** [comment]
I just realized if I merge your work into a separate branch, and then later merge it into `main`, you would not get the proper attribution/credits in the 'Contributors' section on GitHub. So I'm going to have to rethink this.

**Comment by shlomiassaf** [comment]
Don’t worry about it
Just do it

**Comment by mbhoek** [comment]
I think contributions to open source projects should be properly contributed, I really appreciate you spending your valuable time on this. After giving it some thought the solution was pretty simple; enabling pre-release versions of the package allows me to complete PRs without publishing a release version.

Switching your PR back to `main`.

File: SpecFlow.DependencyInjection/DependencyInjectionTestObjectResolver.cs [hunk 1]
```diff
@@ -1,15 +1,50 @@
...
+using System.Reflection;
 using BoDi;
+using Microsoft.Extensions.DependencyInjection;
 using TechTalk.SpecFlow.Infrastructure;
 
 namespace SolidToken.SpecFlow.DependencyInjection
 {
+    /* TODO
+       If SpecFlow will add an "IObjectContainer.IsRegistered(Type type)" method next to the existing "IsRegistered<T>()"
+       We can remove most of the code here!
```
Comment (mbhoek) [inline SpecFlow.DependencyInjection/DependencyInjectionTestObjectResolver.cs:12]: Agreed that would be the cleanest solution, but if I look at `IObjectContainer.Resolve(Type typeToResolve, string name = null);` (in BoDi) it looks like their implementation returns `null` if the type can't be resolved.

Wouldn't it be more pragmatic to default check the `container.Resolve(bindingType)` in line 29-36 and if it returns `null`, then do the call to the `IServiceProvider.GetRequiredService`? 

That would save us from keeping track of these additional `ConcurrentDictionary` and `MethodInfo`s.

File: SpecFlow.DependencyInjection/DependencyInjectionPlugin.cs [hunk 2]
```diff
@@ -12,113 +19,174 @@ namespace SolidToken.SpecFlow.DependencyInjection
...
-            IServiceCollection services)
+        private static void RegisterProxyBindings(IObjectContainer objectContainer, IServiceCollection services)
         {
-            services.AddTransient<FeatureContext>(ctx =>
+            // Required for DI of binding classes that want container injections
+            // While they can (and should) use the method params for injection, we can support it.
+            // Note that in Feature mode, one can't inject "ScenarioContext", this can only be done from method params.
+            
+            // Bases on this: https://docs.specflow.org/projects/specflow/en/latest/Extend/Available-Containers-%26-Registrations.html
+            // Might need to add more...
```
Comment (mbhoek) [inline SpecFlow.DependencyInjection/DependencyInjectionPlugin.cs:137]: I think I want to replace this with the suggested `AddDelegated` extension method mentioned in https://github.com/solidtoken/SpecFlow.DependencyInjection/issues/75#issuecomment-1064906329 but that does not have to be part of this PR now that we have preview release of this package.

**Review by mbhoek — APPROVED** [review]
Looks good to me! :shipit:

**Comment by mbhoek** [comment]
Thanks again for your excellent work @shlomiassaf -- your contribution is now part of the latest preview release.

**Pull request merged**